### ISSUE TITLE
[`len_zero`] suggestion causing compilation fail when macros involved

### ISSUE THREAD

**Issue by J-ZhengLi** [rust-lang/rust-clippy#11992]
### Summary

Somewhat related to PR: #11991 

The issue was that the given suggestion span was inside of a macro, rather than the macro callsite. See [reproduce code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=15d0aad1f097c0f4bddf5b2500b859d7).

I think the lint should skip warnings when the code was expanded from macro.

### Lint Name

len_zero

### Reproducer

I tried this code:

```rust
fn main() {
    macro_rules! len {
        ($seq:ident) => {
            $seq.len()
        }
    }
    
    let s = [1, 2, 3];
    
    if len!(s) == 0 {
        println!("oops");
    }
    (len!(s) == 0).then(|| println!("oops"));
```

I saw this happen:

```
warning: length comparison to zero
  --> src/main.rs:7:13
   |
7  | /             $seq.len()
8  | |         }
9  | |     }
10 | |     
11 | |     let s = [1, 2, 3];
12 | |     
13 | |     if len!(s) == 0 {
   | |________-------____^ help: using `is_empty` is clearer and more explicit: `$seq.is_empty()`
   |          |
   |          in this macro invocation
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#len_zero
...
```

I expected to see this happen:

no warnings, because we might not be able to change the content inside of macros.

### Version

```text
playground: clippy 0.1.74 (2023-12-04 a28077b)
```

### Additional Labels

@rustbot label +I-suggestion-causes-error
